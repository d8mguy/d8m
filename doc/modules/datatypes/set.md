## Set

Module set provides an abstract definition of set and an STMap to go's map type.

    [set(T): lst]

where `lst:list(T)`.

    val set = \typegen(T::entity) {
        val setT = extend tuple(...private...) where {
            method $cvt = \() -> list(T) {...}
            method $in = \(elt: T) -> boolean {...}
            method add = \mod(elt: T) -> nothing {...}
            method add = \mod(other: setT) -> nothing {...}
            method remove = \mod(elt: T) -> nothing {...}
            method remove = \mod(elts: list(T)) -> nothing {...}
            method count = \() -> integer {...}
            method isEmpty = \() -> boolean {...}
            method overlaps = \imp(other: setT) -> boolean {...}
            method $= = \(other: setT) -> boolean {...}
            method subset = \(other: setT) -> boolean {...}
            method intersection = \(other: setT) -> setT {...}
            method $+ = \(other: T) -> setT {...}
            method $+ = \(other: setT) -> setT {...}
        }
    }
