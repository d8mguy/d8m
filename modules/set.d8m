// Module set provides an abstract definition of set and an STMap to go's map type.

export set

val set = \typegen(T::entity) {
	val setT = extend tuple(private x: list(T)) where {
		assert enumerable(x)
		method $cvt = \() -> list(T) { copy(x) }
		method $in = \(elt: T) { elt in x }
		method add = \mod(elt: T) { unless(elt in x) x.pushb(elt) }
		method add = \mod(other: setT) { each(o^other) self.add(o) }
		method remove = \mod(elt: T) { val inx = x.index(elt); if(inx != nil) x.removeIndex(inx) }
		method remove = \mod(elts: list(T)) { each(e^elts) self.remove(e) }
		method lithook = \mod(c: list(T)) { each(x^c) add(x) }
		method count = \() { x.count }
		method isEmpty = \() { x.count == 0 }
		method overlaps = \imp(other: setT) { var rslt = false; each(e^x) if(e in other) { rslt = true; break }; rslt }
		method $== = \(other: setT) -> boolean { other.count == x.count && all(self.{this in other}) }
		// Note: this is self <= other.
		method subset = \(other: setT) { all(self.{this in other}) }
		// Note: intersection could be more efficient, no need for the uniq-check in "rslt.add(o)". But there are tradeoffs.
		method intersection = \imp(other: setT) { var rslt: setT = []; each(o^other) if(o in x) rslt.add(o); rslt }
		method $+ = \imp(other: T) { var rslt: setT = copy(self); rslt.add(other); rslt }
		method $+ = \imp(other: setT) { var rslt: setT = copy(self); each(o^other) rslt.add(o); rslt }
	}
	setT
}
given(T1::entity, y: T1, aLT: list(T1)) {
	STMap {
		source = set(T1)
		target = Gomap(T1, boolean)
		direct count
		y in self => rvalindexWD(self, y, false)
		self.add(y) => self.lvalindex(y, true)
		self.remove(y) => self.delete(y)
		enumerable(self) => self.keys
	}
}
