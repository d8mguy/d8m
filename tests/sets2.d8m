// tests of set enumeration, non-duplication, and equality
import "set" melted
val intset = set(integer)
val s0 = [intset: 1, 17, 3, 5, 3, 22, 18, 1, 33, 8]
val tryIntset = \() {
    println(s0.count) // 8
    val s1 = [intset: s0.{ (this + 2) % 10 }]
    println(s1.count)
    println(s0 == s1)
    println(s1)
}
tryIntset()
