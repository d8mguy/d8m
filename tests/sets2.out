8
6
false
[[3, 9, 5, 7, 4, 0]]
